Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+ * Note that they might have different lengths, so "something" might stand for different
+ * values here. Last-index lists can also be empty.
+ *
+ * The indices array of K=2 contains the indices of the (K=1)-data array at which a new
+ * last-index list starts. If a last-index list is empty, the are repetitive entries in the
+ * indices array.
+ *
+ * The indices array of K=3 contains the indices of the (K=2)-indices array at which the
+ * last-but-two index increases by one or drops to zero. If the last-but-two index is increased
+ * by more than one, there are repetitive entries.
+ *
+ * Etc. etc, up to the indices array at layer K.
+ *
+ * To form such a structure, we typically iterate twice through the pointer-to-pointer-...
+ * array: The first time we get to know how much space to reserve in each layer, then we
+ * allocate it and fill it with data during the second iteration.
+ */
+
+
+template<size_t K, typename Data=su2double, typename Index=unsigned long>
+class NdFlattener;
+
+namespace helpers {
+  template<typename MPI_Allgatherv_type, typename MPI_Datatype_type, typename MPI_Communicator_type>
+  struct NdFlattener_MPI_Environment {
+    MPI_Allgatherv_type MPI_Allgatherv;
+    MPI_Datatype_type mpi_data;
+    MPI_Datatype_type mpi_index;
+    MPI_Communicator_type comm;
+    int rank; int size;
+  };
+
+  /*! \class IndexAccumulator
+   * \brief Data structure holding an offset for the NdFlattener, to provide a []...[]-interface.
+   * \details Derived from IndexAccumulator_Base, specifying the operator[] method:
+   *  - For N==1, the structure has already read all indices but one. So after this method has read the last
+   *    index, return a reference to the data.
+   *  - The case N==2 is much like the case N>3 but an additional function data() should be provided.
+   *  - For N>3, more indices have to be read, return an IndexAccumulator<N-1>.
+   * \tparam N - Number of missing parameters
+   * \tparam Nd_type - Type of the accessed NdFlattener, should be NdFlattener<N,...>
+   */
+  /*! \class IndexAccumulator_Base
+   * \brief Parent class of IndexAccumulator.
+   * \details IndexAccumulator provides the operator[] method.
+   */
+  template<size_t _N, typename _Nd_type>
+  class IndexAccumulator_Base {
+  public:
+    static constexpr size_t N = _N;
+    using Nd_type = _Nd_type;
+
+    /*== Conditional is part of the standard library since C++14, keep it here for C++11 compatibility. ==*/
+    template<bool B, class T, class F> struct conditional { typedef T type; };
+    template<class T, class F> struct conditional<false, T, F> { typedef F type; };
+
+  protected:
+    Nd_type* nd; /*!< \brief The accessed NdFlattener. */
+    const typename Nd_type::Index offset; /*!< \brief Index in the currently accessed layer. */
+    //const typename Nd_type::Index size; /*!< \brief Exclusive upper bound for the next index. */
+    IndexAccumulator_Base(Nd_type* nd, typename Nd_type::Index offset):
```

Comment:
Ok. Now it keeps `IndexAccumulator_Base::nd` by reference instead of pointer. This does also apply to the `local_version` for `NdFlattener` construction via collective communication.

Reply:
Great, less chance for dangling pointer issues.